[clangd] Use `SymbolName` to represent Objective-C selectors

ahoppen opened the pull request:
This is a cleaner design than using identifier and an optional `Selector`. It also allows rename of Objective-C method names if no declaration is at hand and thus no `Selector` instance can be formed. For example, when finding the ranges to rename based on an index that’s not clangd’s built-in index.

CC @DavidGoldman @bnbarham @kadircet

bnbarham commented on clang/lib/Tooling/Refactoring/CMakeLists.txt:
```diff
@@ -13,6 +13,7 @@ add_clang_library(clangToolingRefactoring
   Rename/USRFinder.cpp
   Rename/USRFindingAction.cpp
   Rename/USRLocFinder.cpp
+  SymbolName.cpp
```
Should `SymbolName.cpp` live in `Rename` to match its header?

ahoppen replied:
Goode suggestions 👍🏽

sam-mccall commented on clang/include/clang/Tooling/Refactoring/Rename/SymbolName.h:
```diff
@@ -27,19 +31,45 @@ namespace tooling {
+  /// Creates a \c SymbolName from the given string representation.
+  ///
+  /// For Objective-C symbol names, this splits a selector into multiple pieces
+  /// on `:`. For all other languages the name is used as the symbol name.
```
This looks too much like guesswork. For example in an ObjC++ file (which is our default parse language for *.h), SymbolName(`operator std::string`) will be parsed as the ObjC selector `[" operator std", "", "string"].

If we want to clearly distinguish the different types of names (as clang AST does) then we should avoid implicit conversions like this where it's easy to confuse the encoded/decoded state.

If we're content to mostly use strings and heuristically interpret them as selector names at the edges, that seems OK too - but then this class could just be a "split" function or so.

(This is up to you - I don't mind much what this API looks like unless we end up depending on it)

ahoppen replied:
Turned out that we didn’t actually need this version of the constructor that splits the name based on `:`. I removed it and pass the name pieces directly, which should be more explicit.

sam-mccall commented:
I don't know that this class brings enough value to warrant the dependency - we don't really seem to be simplifying the code, we're mostly just using it as a fancy `vector<string>`.

(For some context: we went to some effort in the past to untangle this from tooling/Refactoring/Rename, and a lot of pieces of clangd make tradeoffs between keeping design simple and handling all the special cases of C, C++, ObjC precisely).

Is the underlying goal here to be able to use `adjustRenameRanges` from outside of clangd? (That's my reading of "finding the ranges to rename based on an index that’s not clangd’s built-in index" - if you were doing this inside clangd, ISTM you'd have a Selector regardless of the index used).
We don't use the selector for anything other than the text chunks it contains, so I think you could just replace `optional<Selector>` with `optional<vector<StringRef>>` there.

(I don't think there are any plans to make use of Selector in other ways, @kadircet @DavidGoldman would know)

sam-mccall commented on clang/include/clang/Tooling/Refactoring/Rename/SymbolName.h:
```diff
@@ -27,19 +31,45 @@ namespace tooling {
 /// //       ^~ string 0 ~~~~~         ^~ string 1 ~~~~~
 /// \endcode
 class SymbolName {
+  llvm::SmallVector<std::string, 1> NamePieces;
```
I'm a little wary of using this class as:

 - it seems like the design has only been informed by objective-C selectors. It's nice to abstract away difficulties of dealing with names, but this doesn't handle many (scope qualifiers, operator names, UDL-names) and it's not clear how/whether it should. If not, I think clangd is better off without the layer of indirection.
 - it's not really clear what you would *do* with this model (name chunks as strings) other than semi-textual rename. If it's a helper class as part of tooling/refactor/rename's API, it might be helpful if it were named/documented as such.

sam-mccall reviewed (changes requested):
As mentioned: unless there's a clear reason, I'd avoid not to add the dependency on Tooling/Refactoring as it doesn't seem to simplify the implementation much.

(My last comment should have been on the review - sorry, I'm still bad at github!)

ahoppen commented:
For some context: When I’m talking about * finding the ranges to rename based on an index that’s not clangd’s built-in index* I meant a request like https://github.com/apple/llvm-project/pull/7973. This allows us to use Apple’s IndexStore to find the locations of symbols to rename and then invoke clangd to get the edits to rename the symbols and deal with the parsing of Objective-C selector pieces. 

Functionality-wise, I would be fine with using a `optional<vector<string>>` instead of `Selector` in `collectRenameIdentifierRanges`. FWIW I disagree that using a `vector<string>` is cleaner than using a dedicated type for it because there’s no type-level information about what the `vector<string>` represents and that `SymbolName` could be made to some day support the cases you mention in https://github.com/llvm/llvm-project/pull/82061#discussion_r1500161008. But I’m new to clangd’s development and will follow your guidance here if you prefer `vector<string>`.

sam-mccall commented:
> For some context: When I’m talking about * finding the ranges to rename based on an index that’s not clangd’s built-in index* I meant a request like [apple#7973](https://github.com/apple/llvm-project/pull/7973).

I see. That makes sense, it's just a bit non-obvious because we don't usually design these pieces as libraries to be consumed outside clangd (either by code calling into clangd's internals, or a modified version of clangd).

I think we usually won't support major design changes to accommodate these, but small ones like this look totally fine.
(There are exceptions, e.g. adding XPC support required significant changes. After these, XPC lives in llvm-project but could just as easily be downstream).

> Functionality-wise, I would be fine with using a `optional<vector<string>>` instead of `Selector` in `collectRenameIdentifierRanges`. FWIW I disagree that using a `vector<string>` is cleaner than using a dedicated type for it because there’s no type-level information about what the `vector<string>` represents

Yeah, "clean" can mean various things, e.g. "clearly communicating intent" here is in tension with "fewer moving parts".
I'd be fine with `struct SelectorName { vector<StringRef> Chunks; }` or `using SelectorName = vector<string>`, more than that is (for my taste) more noise than signal here. I think it should go in `clangd/refactor/Rename.h` though - it's just a part of that API.

ahoppen committed:
[clangd] Use `RenameSymbolName` to represent Objective-C selectors

This is a cleaner design than using identifier and an optional `Selector`. It also allows rename of Objective-C method names if no declaration is at hand and thus no `Selector` instance can be formed. For example, when finding the ranges to rename based on an index that’s not clangd’s built-in index.

ahoppen commented:
Sorry for the super long delay on my side, I got distracted with other tasks over the last few months.

I just rebased the PR and removed the dependency on tooling. Instead, I’m introducing `RenameSymbolName`, which is a typed wrapper around `SmallVector<std::string, 1>` with a few convenience methods added. 

@sam-mccall If you could review the new version of the PR again, I would appreciate it.

ahoppen commented:
@sam-mccall If you could review my latest changes, that would be great.

ahoppen commented:
@DavidGoldman Would you mind reviewing my changes?

DavidGoldman commented on clang-tools-extra/clangd/refactor/Rename.h:
```diff
@@ -35,6 +35,56 @@ struct RenameOptions {
+
+  explicit RenameSymbolName(const DeclarationName &Name);
+
+  // /// Creates a \c SymbolName from the given string representation.
```
This is accidentally double commented, also looks like L63 SymbolName(...) is extraneous.

DavidGoldman commented on clang-tools-extra/clangd/refactor/Rename.cpp:
```diff
@@ -605,12 +607,12 @@ findAllSelectorPieces(llvm::ArrayRef<syntax::Token> Tokens,
       if (PieceCount < NumArgs &&
           isMatchingSelectorName(Tok, Tokens[Index + 1], SM,
-                                 Sel.getNameForSlot(PieceCount))) {
+                                 Name.getNamePieces()[PieceCount])) {
```
nit: can store the pieces as a variable around L600 to make this look a bit cleaner.

DavidGoldman reviewed (changes requested):
Thanks for the fix! Looks good besides some minor nits.

ahoppen commented:
Thanks for the review, @DavidGoldman. I addressed your comments.

DavidGoldman commented on clang-tools-extra/clangd/refactor/Rename.h:
```diff
@@ -35,6 +35,49 @@ struct RenameOptions {
+///
+/// \code
+/// - (void) myMethodNamePiece: (int)x anotherNamePieces:(int)y;
+/// //       ^~ string 0 ~~~~~         ^~ string 1 ~~~~~
```
This is also double indented

ahoppen replied:
Updated. Could you merge the change when CI has passed?
